### ISSUE TITLE
[no-undef] on global namespaces in customized [.d.ts] files from [@typescript-eslint/parser] version [4.x]

### ISSUE THREAD

**Issue by Ares9999** [typescript-eslint/typescript-eslint#3751]
- [ ✔️] I have tried restarting my IDE and the issue persists.
- [ ✔️] I have updated to the latest version of the packages.
- [ ✔️] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Global namespaces are parsed well on ![Awesome Badges](https://img.shields.io/badge/version-3.10.1-blue.svg) , even well with the combination of `eslint-plugin` 4.29.2 and `parser` 3.10.1**

**I have tried using the configurations blow and the issue persists.**

**Repro**
```JSON
"rules": {
    "@typescript-eslint/no-namespace": "off",
    "@typescript-eslint/no-extraneous-class": "off",
    "@typescript-eslint/no-shadow": "off",
}

"parserOptions": {
    "ecmaVersion": 2019,
    "sourceType": "module",
    "extraFileExtensions": [
        "*.d.ts"
    ]
},
```
**Example codes**
```TS
// my sc.d.ts file inside one of the project's directory
declare namespace sc {
    type MyType: string;
}

// my a.ts file
export default abstract class A {
    // here comes the `no-undef` error on `sc`.
    // the `sc` can still be linked by `ctrl+click` in VsCode
    // no error on version `3.10.1`
    public static init(args: sc.MyType) : void {
        return null;
    }
}
```
**Expected Result**
- Parses the namespaces in customized d.ts file like `3.10.1`
- Or removes the error
- Or sets up a new configuration to let us ignore this error by not setting `no-undef` to `off` in eslinc.json.

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.x` |
| `@typescript-eslint/parser` | `4.x` |
| `TypeScript`                | `4.3.5` |
| `ESLint`                    | `7.32.0` |
| `node`                      | `v12.18.4` |

**Comment by bradzacher** [comment]
> [ ✔️] I have read the FAQ and my problem is not listed.

Did you really read it...?

https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-get-errors-from-the-no-undef-rule-about-global-variables-not-being-defined-even-though-there-are-no-typescript-errors

**Comment by Ares9999** [comment]
Lol, sorry!
I did saw this undef FAQ today but only paid attention to the quoted part.
Cuz I remember I might've tried the global configuration in ESlint config on the first 4.x versions and it didn't work. Since then I just update and see if the undef remains and roll back.


Thanks a lot for your time.






  
[ ✔️] I have read the FAQ and my problem is not listed.
  
Did you really read it...?
 
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-get-errors-from-the-no-undef-rule-about-global-variables-not-being-defined-even-though-there-are-no-typescript-errors
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
